### PR TITLE
Greatly reduce GCC build logs

### LINKED ISSUE #80: Log often shows lots of irrelevant build messages
Consider e.g. [this](https://github.com/rust-lang/rust/pull/122471#issuecomment-1996778138): the actually relevant error is 
```
2024-03-14T08:02:45.9904842Z failures:
2024-03-14T08:02:45.9905007Z 
2024-03-14T08:02:45.9906975Z ---- [ui] tests/ui/generics/post_monomorphization_error_backtrace.rs stdout ----
2024-03-14T08:02:45.9907581Z diff of stderr:
2024-03-14T08:02:45.9907742Z 
2024-03-14T08:02:45.9907827Z 6	   |
2024-03-14T08:02:45.9908550Z 7	   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)
2024-03-14T08:02:45.9909539Z 8	
2024-03-14T08:02:45.9909782Z +	note: erroneous constant encountered
2024-03-14T08:02:45.9910305Z +	  --> $DIR/post_monomorphization_error_backtrace.rs:14:5
2024-03-14T08:02:45.9910713Z +	   |
2024-03-14T08:02:45.9914947Z +	LL |     F::<T>::V;
2024-03-14T08:02:45.9915253Z +	   |     ^^^^^^^^^
2024-03-14T08:02:45.9915626Z +	
2024-03-14T08:02:45.9916110Z 9	note: the above error was encountered while instantiating `fn assert_zst::<u32>`
2024-03-14T08:02:45.9916904Z 10	  --> $DIR/post_monomorphization_error_backtrace.rs:18:5
2024-03-14T08:02:45.9917516Z 11	   |
2024-03-14T08:02:45.9917652Z 
2024-03-14T08:02:45.9918749Z 19	   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: std::mem::size_of::<T>() == 0', $DIR/post_monomorphization_error_backtrace.rs:6:23
2024-03-14T08:02:45.9919579Z 20	   |
2024-03-14T08:02:45.9920369Z 21	   = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)
2024-03-14T08:02:45.9921439Z +	
2024-03-14T08:02:45.9921823Z +	note: erroneous constant encountered
2024-03-14T08:02:45.9922627Z +	  --> $DIR/post_monomorphization_error_backtrace.rs:14:5
2024-03-14T08:02:45.9923287Z +	   |
2024-03-14T08:02:45.9923514Z +	LL |     F::<T>::V;
2024-03-14T08:02:45.9923773Z +	   |     ^^^^^^^^^
2024-03-14T08:02:45.9924014Z +	   |
2024-03-14T08:02:45.9924491Z +	   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
2024-03-14T08:02:45.9925018Z 22	
2024-03-14T08:02:45.9925545Z 23	note: the above error was encountered while instantiating `fn assert_zst::<i32>`
2024-03-14T08:02:45.9926259Z 24	  --> $DIR/post_monomorphization_error_backtrace.rs:18:5
2024-03-14T08:02:45.9926595Z 
2024-03-14T08:02:45.9926625Z 
2024-03-14T08:02:45.9926804Z The actual stderr differed from the expected stderr.
2024-03-14T08:02:45.9928017Z Actual stderr saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generics/post_monomorphization_error_backtrace/post_monomorphization_error_backtrace.stderr
2024-03-14T08:02:45.9929234Z To update references, rerun the tests and pass the `--bless` flag
2024-03-14T08:02:45.9930149Z To only update this specific test, also pass `--test-args generics/post_monomorphization_error_backtrace.rs`
2024-03-14T08:02:45.9930729Z 
2024-03-14T08:02:45.9937039Z error: 1 errors occurred comparing output.
2024-03-14T08:02:45.9937479Z status: exit status: 1
2024-03-14T08:02:45.9942691Z command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generics/post_monomorphization_error_backtrace" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/generics/post_monomorphization_error_backtrace/auxiliary"
2024-03-14T08:02:45.9947953Z stdout: none
2024-03-14T08:02:45.9948292Z --- stderr -------------------------------
2024-03-14T08:02:45.9948785Z error[E0080]: evaluation of `assert_zst::F::<u32>::V` failed
2024-03-14T08:02:45.9963235Z ##[error]  --> /checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs:6:23
2024-03-14T08:02:45.9970135Z    |
2024-03-14T08:02:45.9970513Z LL |         const V: () = assert!(std::mem::size_of::<T>() == 0);
2024-03-14T08:02:45.9971762Z    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: std::mem::size_of::<T>() == 0', /checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs:6:23
2024-03-14T08:02:45.9972581Z    |
2024-03-14T08:02:45.9973227Z    = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)
2024-03-14T08:02:45.9973817Z 
2024-03-14T08:02:45.9973938Z note: erroneous constant encountered
2024-03-14T08:02:45.9974545Z   --> /checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs:14:5
2024-03-14T08:02:45.9975060Z    |
2024-03-14T08:02:45.9975268Z LL |     F::<T>::V;
2024-03-14T08:02:45.9975509Z    |     ^^^^^^^^^
2024-03-14T08:02:45.9975658Z 
2024-03-14T08:02:45.9975943Z note: the above error was encountered while instantiating `fn assert_zst::<u32>`
2024-03-14T08:02:45.9976862Z   --> /checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs:18:5
2024-03-14T08:02:45.9977415Z    |
2024-03-14T08:02:45.9977624Z LL |     assert_zst::<U>()
2024-03-14T08:02:45.9977884Z    |     ^^^^^^^^^^^^^^^^^
2024-03-14T08:02:45.9978056Z 
2024-03-14T08:02:45.9978295Z error[E0080]: evaluation of `assert_zst::F::<i32>::V` failed
2024-03-14T08:02:45.9979606Z ##[error]  --> /checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs:6:23
2024-03-14T08:02:45.9980659Z    |
2024-03-14T08:02:45.9980974Z LL |         const V: () = assert!(std::mem::size_of::<T>() == 0);
2024-03-14T08:02:45.9982286Z    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'assertion failed: std::mem::size_of::<T>() == 0', /checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs:6:23
2024-03-14T08:02:45.9983180Z    |
2024-03-14T08:02:45.9983843Z    = note: this error originates in the macro `assert` (in Nightly builds, run with -Z macro-backtrace for more info)
2024-03-14T08:02:45.9984429Z 
2024-03-14T08:02:45.9984550Z note: erroneous constant encountered
2024-03-14T08:02:45.9985164Z   --> /checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs:14:5
2024-03-14T08:02:45.9985677Z    |
2024-03-14T08:02:45.9985886Z LL |     F::<T>::V;
2024-03-14T08:02:45.9986124Z    |     ^^^^^^^^^
2024-03-14T08:02:45.9986348Z    |
2024-03-14T08:02:45.9986800Z    = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`
2024-03-14T08:02:45.9987219Z 
2024-03-14T08:02:45.9987520Z note: the above error was encountered while instantiating `fn assert_zst::<i32>`
2024-03-14T08:02:45.9988316Z   --> /checkout/tests/ui/generics/post_monomorphization_error_backtrace.rs:18:5
2024-03-14T08:02:45.9988913Z    |
2024-03-14T08:02:45.9989132Z LL |     assert_zst::<U>()
2024-03-14T08:02:45.9989414Z    |     ^^^^^^^^^^^^^^^^^
2024-03-14T08:02:45.9989588Z 
2024-03-14T08:02:45.9989752Z error: aborting due to 2 previous errors
2024-03-14T08:02:45.9990019Z 
2024-03-14T08:02:45.9990324Z For more information about this error, try `rustc --explain E0080`.
2024-03-14T08:02:45.9990828Z ------------------------------------------
2024-03-14T08:02:45.9991065Z 
2024-03-14T08:02:45.9991071Z 
2024-03-14T08:02:45.9991076Z 
2024-03-14T08:02:45.9991163Z failures:
2024-03-14T08:02:45.9991514Z     [ui] tests/ui/generics/post_monomorphization_error_backtrace.rs
```
However, the bot shows huge amounts of irrelevant noise above that, and it cuts off the interesting part at the end that lists the failed testcases.

I don't know whether the right answer is to make CI logging less verbose or the log analysis better. I think all that noise in the CI logs is from the GCC backend tests? At least it seems to be about building GCC. I have no idea why those tests are so extremely noisy when compared with everything else we have on CI. Looing at the [raw logs](https://productionresultssa18.blob.core.windows.net/actions-results/588b5472-4fc1-436a-9ee5-e498ca25ce80/workflow-job-run-a64ed217-e4bb-5680-9c18-8f3ddeed4bfc/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-03-14T08%3A28%3A54Z&sig=uSSuoyO8DsKMWmXE2wGHjrgBHOQCXlQeX5AHeI2Uls0%3D&sp=r&spr=https&sr=b&st=2024-03-14T08%3A18%3A49Z&sv=2021-12-02), it seems like the GCC build is responsible for >80% of all log output! (That's from the first mention of "gcc-src" until the last.) Having all that noise in the log makes it harder to find the real problems.

cc @antoyo, @GuillaumeGomez

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/rust-lang/rust-log-analyzer/issues/80.

Based on [makefile documentation](https://www.gnu.org/software/make/manual/html_node/Options-Summary.html#index-_002d_002dquiet-1) and [configure documentation](https://www.gnu.org/software/autoconf/manual/autoconf-2.68/html_node/configure-Invocation.html).

cc @RalfJung @antoyo

### CODE REVIEW

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh [hunk 1]
```diff
@@ -21,8 +21,9 @@ cd ../gcc-build
     --disable-bootstrap \
     --disable-multilib \
     --prefix=$(pwd)/../gcc-install
```
Comment (RalfJung) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:23]: Doesn't this need a backslash? Otherwise you're calling a new command `--quiet`...

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh]: Arf, thanks!

**Review by antoyo — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
The `--quiet` option isn't enough apparently... Let's do the good old `> /dev/null`.

**Comment by RalfJung** [comment]
That will be annoying if it ever fails though...
Is there an upstream issue tracking "please provide a build option that outputs less than 100 lines on success"?

**Comment by antoyo** [comment]
> That will be annoying if it ever fails though... Is there an upstream issue tracking "please provide a build option that outputs less than 100 lines on success"?

The errors are printed on stderr, so it's not an issue imo.

**Comment by RalfJung** [comment]
In my experience the output often becomes hard to follow when throwing away just stdout, as nobody tests that.

But sure, we can give it a shot.

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh [hunk 2]
```diff
@@ -20,9 +20,10 @@ cd ../gcc-build
     --enable-checking=release \
     --disable-bootstrap \
     --disable-multilib \
-    --prefix=$(pwd)/../gcc-install
-make
-make install
+    --prefix=$(pwd)/../gcc-install > /dev/null
+
+make > /dev/null
+make install > /dev/null
```
Comment (RalfJung) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:26]: Are *all* of these steps generating thousands of lines?

Also please add a comment explaining why we're taking such drastic measures.

Comment (antoyo) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:26]: The configure (if we added back `--quiet`) might give only a dozen lines, but `make` will definitely generate thousands of lines (even with `--quiet`). For `make install`, it might be only a couple hundred lines with `--quiet`.

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:26]: I still think it's way too much (for all commands). So better keep it this way as it's not relevant to the build.

I'll add comments.

**Comment by TimNN** [comment]
I don't know if it's easily installable in whatever OS the CI Docker container uses, but there's a tool called `chronic` that

> runs a command, and arranges for its standard out and standard error to only be
>       displayed if the command fails (exits nonzero or crashes).  If the command succeeds, any
 >     extraneous output will be hidden.

Which seems like it would be useful here.

https://manpages.ubuntu.com/manpages/noble/man1/chronic.1.html

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh (hunk 2 above)
Comment (RalfJung) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:29]: > The configure (if we added back --quiet) might give only a dozen lines,

Seems reasonable to just keep that visible then.

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:29]: Ok, done.

Comment (antoyo) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:29]: You forgot to add back `--quiet` to configure, no?

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:29]: Not sure it changes anything but let's put it back.

**Comment by klensy** [comment]
Well, you can tune log analyzer to skip these build logs instead turning off logs.

**Comment by RalfJung** [comment]
Sometimes one has to look at the logs though. I don't think it is reasonable to have 80% of them filled with the GCC build when that's a tiny part of what we are testing.

**Comment by klensy** [comment]
It wasn't an issue before, so there something changed in pipeline?

**Comment by RalfJung** [comment]
It's not been that long since GCC was added to the build. I think it was a problem ever since, or do you have evidence to the contrary?

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh [hunk 3]
```diff
@@ -20,9 +20,14 @@ cd ../gcc-build
...
-make
-make install
+    --prefix=$(pwd)/../gcc-install \
+    --quiet
+
+# We redirect stdout because the GCC build scripts generate thousands of lines
+# of output. Unfortunately, even when using `--quiet` option for all commands,
+# the output still thousands of lines of output, forcing us to use this solution.
+make > /dev/null
+make install > /dev/null
```
Comment (Mark-Simulacrum) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:30]: Please use hide_output from src/ci/docker/scripts/shared.sh (feel free to grep for other usages). That makes sure there's some progress indication and we do the right thing on failure.

**Comment by RalfJung** [comment]
Btw is it expected that the GCC build takes about 30 minutes? (If I understand [this log](https://productionresultssa19.blob.core.windows.net/actions-results/949eab71-f9f1-4390-8a18-61665488acb3/workflow-job-run-a64ed217-e4bb-5680-9c18-8f3ddeed4bfc/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-03-15T06%3A44%3A26Z&sig=6iqHyHG2WZBjYeyDmihDDTH7WvdlYVXJEM1vzaW0BX4%3D&sp=r&spr=https&sr=b&st=2024-03-15T06%3A34%3A21Z&sv=2021-12-02) correctly.) That's 33% of the entire CI job! And in particular this happens *before* the rustc test suite is run, greatly increasing the latency until regular test failures are discovered.

**Comment by RalfJung** [comment]
Also why do both the gnu-llvm-16 and the gnu-tools runner build GCC...?

**Comment by GuillaumeGomez** [comment]
> Also why do both the gnu-llvm-16 and the gnu-tools runner build GCC...?

It is run it 3 CIs: gnu-tools, llvm-16 and llvm-17. When I made the last sync of `rustc_codegen_gcc`, I removed the installations of libgccjit through the system and replaced it with the build of GCC (you can see that in the last commits of #122042, like https://github.com/rust-lang/rust/pull/122042/commits/c5c6729c88b6d00798f46d5c7397e854775d0923).

If people agree on reducing the number of CIs building GCC, I can remove some of them.

**Comment by GuillaumeGomez** [comment]
> Btw is it expected that the GCC build takes about 30 minutes? (If I understand [this log](https://productionresultssa19.blob.core.windows.net/actions-results/949eab71-f9f1-4390-8a18-61665488acb3/workflow-job-run-a64ed217-e4bb-5680-9c18-8f3ddeed4bfc/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-03-15T06%3A44%3A26Z&sig=6iqHyHG2WZBjYeyDmihDDTH7WvdlYVXJEM1vzaW0BX4%3D&sp=r&spr=https&sr=b&st=2024-03-15T06%3A34%3A21Z&sv=2021-12-02) correctly.) That's 30% of the entire CI job! And in particular this happens _before_ the rustc test suite is run, greatly increasing the latency until regular test failures are discovered.

For now it's built using only one thread. I don't know if I can get the number of CPUs available in the CI and pass it to the `make` scripts. That would likely greatly improve the time it takes in the CI.

**Comment by Mark-Simulacrum** [comment]
https://github.com/rust-lang/rust/blob/master/src/ci/docker/host-x86_64/dist-x86_64-linux/build-clang.sh#L42 is an example. It also demonstrates the use of hide_output.

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh [hunk 4]
```diff
@@ -14,15 +16,17 @@ mkdir ../gcc-build ../gcc-install
...
     --enable-languages=jit \
     --enable-checking=release \
     --disable-bootstrap \
     --disable-multilib \
-    --prefix=$(pwd)/../gcc-install
-make
-make install
+    --prefix=$(pwd)/../gcc-install \
+
+hide_output make -j$(nproc)
```
Comment (RalfJung) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:28]: This seems to speed up the build by a factor of 4.^^ It was 30min before, now it's...
```
2024-03-15T12:47:03.8776557Z #14 DONE 446.8s
```

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh [hunk 5]
```diff
@@ -1,7 +1,9 @@
-#!/bin/sh
+#!/usr/bin/env bash
 
 set -ex
 
+source /tmp/shared.sh
```
Comment (RalfJung) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:5]: Why didn't you put this in the same folder as the build-gccjit script?

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh (hunk 4 above)
Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:28]: Nice. :)

File: src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh (hunk 5 above)
Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/dist-x86_64-linux/build-gccjit.sh:5]: Mostly because I was first trying to make it work. Cleanup time!

**Comment by GuillaumeGomez** [comment]
Clean up done. Nice speed up in any case. :)
